### PR TITLE
`gr.Dropdown()` now supports values with arbitrary characters and doesn't clear value when re-focused

### PR DESCRIPTION
**Pull request by abidlabs**
Fixes two separate issues in `gr.Dropdown()`:

**1. The value of the `gr.Dropdown()` was getting cleared when clicked.** 

The reason that we were doing this was because we wanted users to be able to easily see the full list of choices when they focused on a `gr.Dropdown` and had already previously selected a choice. Ordinarily, the list of choices would get filtered by the value of the `gr.Dropdown()` so you needed to clear the value to see the full list. However, it was a little weird to see the value of the `gr.Dropdown()` clearing on focus and was especially problematic if `allow_custom_value` was set to `True` as it made the value not really editable after you had left the focus on the `gr.Dropdown()`. 

Instead, I tweaked the behavior of the `gr.Dropdown()` so that the full list of choices is always shown immediately on focus without needing to clear the value of the `gr.Dropdown()`. Once you start typing, the list of choices gets filtered again.

Test with:

```py
import gradio as gr

with gr.Blocks() as demo:
    gr.Dropdown(choices=["America", "Azerbaijan", "Bahrain"], allow_custom_value=True)    
    gr.Dropdown(choices=["America", "Azerbaijan", "Bahrain"], allow_custom_value=False)
    
demo.launch()
```

 
Closes: #4555. 

**2. The `gr.Dropdown()` was not working properly if values with whitespaces and other special characters were used.**

Because we were using 

```js
document.querySelector(`li[data-value="${value}"]`
```

we couldn't handle certain characters. Now we avoid using `document.querySelector()`. Test with:

```py
import gradio as gr

with gr.Blocks() as demo:
    gr.Dropdown(choices=["I love you", "Can't believe it", "Here's a \" for you", "Normal"])
    
demo.launch()
```

Closes: #4503, Closes: #4292

Other small changes:
* I reordered the parameters to make `allow_custom_value` more discoverable in the [docs as discussed internally](https://huggingface.slack.com/archives/C02QZLG8GMN/p1690426170569499?thread_ts=1690369398.426209&cid=C02QZLG8GMN)
* Fixed some small linting issues

Kudos to the function tests which caught a subtle bug in my original implementation!

cc @abhishekkrthakur @yvrjsharma

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
 	"compilerOptions": {
 		"moduleResolution": "node",
 		"module": "es2020",
-		"lib": ["es2020", "DOM"],
+		"lib": ["es2020", "DOM", "dom.iterable"],
```
Comment (abidlabs) [inline tsconfig.json:5]: This is needed so that TS doesn't complain about this line:
```
				for (const element of elements) {
```

**Comment by abidlabs** [comment]
Thank you for the review @hannahblair!

**Comment by abhishekkrthakur** [comment]
will this be released soon?

**Comment by abidlabs** [comment]
Probably later this week. But in the meantime, you can install using the link above: pip install https://gradio-builds.s3.amazonaws.com/2b70de43cc1731a843392fbbbddff97ef5e10d18/gradio-3.39.0-py3-none-any.whl

**Pull request merged**